presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -366,15 +397,48 @@ public void addInput(Page page)
         inputProcessed = true;
 
         initializeAggregationBuilderIfNeeded();
+        if (preGroupedHashStrategy.isPresent()) {
+            // Running in segmented aggregation mode.
+            int lastRowInPage = page.getPositionCount() - 1;
+            if (currentSegment == null
+                    || preGroupedHashStrategy.get().rowEqualsRow(0, currentSegment.extractChannels(preGroupedChannels), lastRowInPage, page.extractChannels(preGroupedChannels))) {
+                // If this is the first page or the current segment might have more data in the incoming pages, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
+            }
+            else if (preGroupedHashStrategy.get().rowEqualsRow(0, currentSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
+                // If the current page starts with a new segment, flush before processing it.
+                remainingPage = page;
+            }
+            else {
+                // If the current segment ends in the current page, flush it with all the segments (if exist) except the last segment of the current page.
+                int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
+                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
+                remainingPage = page.getRegion(0, lastRowInPage - lastSegmentStart + 1);
+            }
+            // Record the last segment.
+            currentSegment = page.getRegion(lastRowInPage, 1);
```

Comment:
Here I understand the end to end logic is correct, but if you look at logic details, it's a bit inconsistent. 

For example, for the first page, you always treat it specially. So even if we can do flush for the first page, you won't. But for the rest of the pages, you will do flush. 


I haven't verified yet, but what do you think about changing it to something like:
 ```
            int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
            if (lastSegmentStart == 0) {
                if (currentSegment == null || preGroupedHashStrategy.get().rowEqualsRow(0, currentSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
                    unfinishedWork = aggregationBuilder.processPage(page);
                }
                else {
                    remainingPage = page;
                }
            }
            else {
                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
                remainingPage = page.getRegion(0, lastRowInPage - lastSegmentStart + 1);
            }
```

Reply:
Not sure if I fully understand.

For the first page, I think it needs to be treated specially because if there is no page before it, it is guaranteed that there is no data waiting to be flushed. The operator can just go ahead and process the whole page.

If you mean the first row is treated specially, this is a small optimization I add - if the page starts a new segment, instead of scanning the whole page backwards row by row, we can just take a quick look at the first row. But let me know if it is over-engineered.

Reply:
For example, if we have page 1 [1, 1, 1......1, 2] and page 2 [2, 2, 2...2, 3]; 
For page1, you won't do flush for value 1 even though we already have at least one segment ready to be flushed
However for page2, you would definitely flush for value 2

In your `shouldFlush` function
"// - running in segmented aggregation mode and at least one segment has been fully processed"

Reply:
I see you point. Will fix the inconsistency.

Reply:
I just simply the logic, let me know if you think it makes sense

```
        // Segmented aggregation
        if (preGroupedHashStrategy.isPresent()) {
            int lastRowInPage = page.getPositionCount() - 1;
            int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
            // the whole page is one segment and we don't know where this segment ends
            if (lastSegmentStart == 0) {
                unfinishedWork = aggregationBuilder.processPage(page);
            }
            else {
                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
                remainingPageForSegmentedAggregation = page.getRegion(0, lastRowInPage - lastSegmentStart + 1);
            }

            // Record the first unfinished segment
            firstUnfinishedSegment = page.getRegion(lastRowInPage, 1);
        }
        // Normal aggregation
        else {
            unfinishedWork = aggregationBuilder.processPage(page);
        }
```

Reply:
This logic doesn't cover the case like: `page 1 [1, 1, 1......1, 1], page 2 [2, 2, 2...2, 2], page 3 [2, 2, 2...2, 2]` - it will never flush.
Also in the above case, after the first page got processed and the `firstUnfinishedSegment` became `1`, when we reach page2, we'll scan the whole page 2 row by row and then realize the whole page is a new segment because `lastSegmentStart == 0`. And we'll do `page.getRegion(0, lastRowInPage - lastSegmentStart + 1)` which basically copied the whole page to `remainingPageForSegmentedAggregation` - (there is another for-loop inside). Instead, we can quickly check the first row if it's a new segment and avoid the row-by-row scan/copy. WDYT?

Reply:
Good catch, then I guess the previous one I provided would work?